File: src/api/queries/n_nearest/NNearestQueryController.cs
using System;
using System.Runtime.InteropServices;
using Microsoft.AspNetCore.Http;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using DVAN.Routing;
using System.Threading.Tasks;
using DVAN.Accessibility;
using DVAN.Population;
using NetTopologySuite.Geometries;

namespace DVAN.API
{
    [ApiController]
    [Route("/v1/queries/n_nearest")]
    public class NNearestQueryController : ControllerBase
    {
        static Dictionary<Guid, NNearestQuerySession> sessions = new Dictionary<Guid, NNearestQuerySession>();

        ILogger logger;

        public NNearestQueryController(ILogger<NNearestQueryController> logger)
        {
            this.logger = logger;
        }

        /// <summary>
        /// Calculates nnearest query.
        /// </summary>
        [HttpPost]
        [ProducesResponseType(200, Type = typeof(NNearestQueryResponse))]
        [ProducesResponseType(400, Type = typeof(ErrorResponse))]
        public async Task<IActionResult> calcQuery([FromBody] NNearestQueryRequest request)
        {
            IKNNTable table;
            IPopulationView? view;
            Guid session_id;
            if (request.session_id != null) {
                session_id = request.session_id.Value;
                if (!sessions.ContainsKey(session_id)) {
                    return BadRequest(new ErrorResponse("n_nearest", "no active session found"));
                }
                var session = sessions[session_id];
                table = session.table;
                view = session.population_view;
            }
            else {
                view = PopulationManager.getPopulationView(request.population);
                if (view == null) {
                    return BadRequest(new ErrorResponse("accessibility/gravity/grid", "failed to get population-view, parameters are invalid"));
                }
                IRoutingProvider provider = RoutingManager.getRoutingProvider(request.routing);

                table = await provider.requestKNearest(view, request.facility_locations, request.ranges, request.facility_count, "isochrones");

                session_id = Guid.NewGuid();
                sessions[session_id] = new NNearestQuerySession {
                    id = session_id,
                    table = table,
                    population_view = view
                };
            }

            var results = NNearestQuery.computeQuery(request.facility_values, table, request.compute_type, request.facility_count);

            return Ok(new NNearestQueryResponse {
                result = results,
                session_id = session_id,
            });
        }
    }
}

File: src/main/java/org/tud/oas/api/queries/n_nearest/NNearestQueryController.java
package org.tud.oas.api.queries.n_nearest;

import org.tud.oas.demand.IDemandView;
import org.tud.oas.responses.ErrorResponse;
import org.tud.oas.routing.IKNNTable;
import org.tud.oas.routing.IRoutingProvider;
import org.tud.oas.routing.RoutingOptions;
import org.tud.oas.services.DecayService;
import org.tud.oas.services.DemandService;
import org.tud.oas.services.RoutingService;
import org.tud.oas.services.SupplyService;
import org.tud.oas.supply.ISupplyView;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@RestController
@RequestMapping("/v1/queries/n_nearest")
public class NNearestQueryController {
    static Map<UUID, NNearestQuerySession> sessions = new ConcurrentHashMap<>();
    private final Logger logger = LoggerFactory.getLogger(NNearestQueryController.class);

    private RoutingService routing_service;
    private DemandService demand_service;
    private SupplyService supply_service;

    @Autowired
    public NNearestQueryController(RoutingService routing, DemandService demand, SupplyService supply) {
        this.routing_service = routing;
        this.demand_service = demand;
        this.supply_service = supply;
    }

    @Operation(description = """
            Calculates nnearest query.
            """)
    @ApiResponse(responseCode = "200", description = "Standard response for successfully processed requests.", content = {
            @Content(mediaType = "application/json", schema = @Schema(implementation = NNearestQueryResponse.class))
    })
    @ApiResponse(responseCode = "400", description = "The request is incorrect and therefore can not be processed.", content = {
            @Content(mediaType = "application/json", schema = @Schema(implementation = ErrorResponse.class))
    })
    @PostMapping
    public ResponseEntity<?> calcQuery(@RequestBody NNearestQueryRequest request) {
        IKNNTable table;
        IDemandView demand_view;
        UUID sessionId;
        ISupplyView supply_view = supply_service.getSupplyView(request.supply);
        if (supply_view == null) {
            return ResponseEntity.badRequest().body(new ErrorResponse("queries/n_nearest",
                    "failed to get supply-view, parameters are invalid"));
        }

        if (request.session_id != null) {
            sessionId = request.session_id;
            if (!sessions.containsKey(sessionId)) {
                return ResponseEntity.badRequest().body(new ErrorResponse("n_nearest", "no active session found"));
            }
            NNearestQuerySession session = sessions.get(sessionId);
            table = session.table;
            demand_view = session.demand_view;
        } else {
            demand_view = demand_service.getDemandView(request.demand);
            if (demand_view == null) {
                return ResponseEntity.badRequest().body(new ErrorResponse("accessibility/gravity/grid",
                        "failed to get population-view, parameters are invalid"));
            }

            IRoutingProvider provider = routing_service.getRoutingProvider(request.routing);
            table = provider.requestKNearest(demand_view, supply_view, request.facility_count,
                    new RoutingOptions("isochrones", request.ranges));

            sessionId = UUID.randomUUID();
            sessions.put(sessionId, new NNearestQuerySession(sessionId, demand_view, table));
        }

        float[] results = NNearestQuery.computeQuery(supply_view, table, request.compute_type,
                request.facility_count);

        return ResponseEntity.ok(new NNearestQueryResponse(results, sessionId));
    }
}
